### PR TITLE
Obstacle avoidance

### LINKED ISSUE #444: Implement a Theta* path planner
### Description of the task

Implement a Theta* path planner based on the [DOSL library](https://github.com/subh83/DOSL)

### Outline the task in more detail

Implement a Theta* path planner based on the [DOSL library](https://github.com/subh83/DOSL)

Depends on #443, #440, #441, #442 

### Acceptance criteria

- [ ] Theta* path planner

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

This PR is a superset of #648 with additional changes by @jonl112 	to allow generic obstacles to be passed to the navigator, which gives the base framework to allow high-level logic to avoid certain areas based on the current gamestate.

### Testing Done

Existing unit tests and manual verification, both in grsim and on-field

### Resolved Issues

resolves #444

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — DISMISSED** [review]
LGTM!

File: src/thunderbots/software/ai/navigator/obstacle/obstacle.h [hunk 1]
```diff
@@ -3,6 +3,8 @@
  */
 #pragma once
 
+#include <ai/primitive/primitive.h>
```
Comment (jordanzliu) [inline src/thunderbots/software/ai/navigator/obstacle/obstacle.h:6]: nit: should be `#include "ai/primitive/primitive.h"`

File: src/thunderbots/software/ai/navigator/obstacle/obstacle.h [hunk 2]
```diff
@@ -14,8 +16,35 @@
...
+     *
+     * @param start             start position
+     * @param destination       destination position
+     * @param initial_speed     initial speed of the robot
+     * @param width_scaling     multiplicatively scales the width of obstacle
+     * @param length_scaling    multiplicatively scales the length
+     *
+     * @return a rectangular Polygon to represent the boundary around the obstacle
+     */
+    static Obstacle createVelocityObstacleWithScalingParams(Point start, Point end,
```
Comment (jordanzliu) [inline src/thunderbots/software/ai/navigator/obstacle/obstacle.h:43]: IMO this function name is too verbose, we can trim it down to something like `createRobotVelocityObstacle`.

File: src/thunderbots/software/ai/navigator/obstacle/obstacle.h [hunk 3]
```diff
@@ -65,12 +94,15 @@ class Obstacle
...
+    static Obstacle createBallObstacle(const Ball& ball,
+                                       double additional_radius_cushion_buffer,
+                                       double additional_velocity_cushion_buffer);
 
     const Polygon& getBoundaryPolygon() const;
 
+
    private:
-    Obstacle(Polygon polygon);
     static Obstacle createRobotObstacleFromPositionAndRadiusAndVelocity(
```
Comment (jordanzliu) [inline src/thunderbots/software/ai/navigator/obstacle/obstacle.h:106]: function name too verbose, we can at least take out the `And`s

File: src/thunderbots/software/ai/intent/move_intent.h [hunk 4]
```diff
@@ -65,9 +66,25 @@ class MoveIntent : public Intent, public MovePrimitive
...
+     * @return additional obstacles for this move intent
+     */
+    std::vector<Obstacle> getAdditionalObstacles() const;
+
+    /*
+     * Adds an additional obstaclesfor this move intent
+     *
+     * @param an additional obstacle for this move intent
+     */
+    void getAdditionalObstacles(Obstacle o);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/ai/intent/move_intent.h:81]: Changing this in another pr that should go in this evening.

**Comment by garethellis0** [comment]
Looks like CI segfaulted on navigator tests:
```
[       OK ] PathPlanningNavigatorTest.convert_direct_velocity_intent_to_direct_velocity_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_direct_wheels_intent_to_direct_wheels_primitive
[       OK ] PathPlanningNavigatorTest.convert_direct_wheels_intent_to_direct_wheels_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_dribble_intent_to_dribble_primitive
[       OK ] PathPlanningNavigatorTest.convert_dribble_intent_to_dribble_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_kick_intent_to_kick_primitive
[       OK ] PathPlanningNavigatorTest.convert_kick_intent_to_kick_primitive (1 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_movespin_intent_to_movespin_primitive
[       OK ] PathPlanningNavigatorTest.convert_movespin_intent_to_movespin_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_pivot_intent_to_pivot_primitive
[       OK ] PathPlanningNavigatorTest.convert_pivot_intent_to_pivot_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_stop_intent_to_stop_primitive
[       OK ] PathPlanningNavigatorTest.convert_stop_intent_to_stop_primitive (0 ms)
[ RUN      ] PathPlanningNavigatorTest.convert_multiple_intents_to_primitives
Segmentation fault (core dumped)
Cannot find results, writing failure results to '/root/Software/build/test_results/thunderbots/MISSING-gtest-navigator_test.xml'
-- run_tests.py: execute commands
  /root/Software/devel/lib/thunderbots/navigator_test --gtest_output=xml:/root/Software/build/test_results/thunderbots/gtest-navigator_test.xml
-- run_tests.py: verify result "/root/Software/build/test_results/thunderbots/gtest-navigator_test.xml"
[100%] Built target _run_tests_thunderbots_gtest_navigator_test
Scanning dependencies of target _run_tests_thunderbots_gtest_passing_test
```

**Review by Thunderbots — APPROVED** [review]

**Pull request merged**